es6-shim.js

```diff
@@ -291,6 +291,13 @@
       return x != null && Object(x) === x;
     },
 
+    TypeIsNotObject: function (val) {
+      // This is faster when TypeIsObject may commonly return false.
+      return val === void 0 || val === null || val === true || val === false ||
```

Comment:
I think the key to the optimization is that `undefined`, `null`, `true`, `false` etc are all actually constant pointers which can be compared directly without dereferencing.  Similarly `number` is a tagged value (a tagged NaN) so it can also be a direct comparison.   That makes those tests extremely fast if they return true.  We only have to actually dereference for `string` and `symbol` (and I might be wrong about those, likely there are some optimizations for string) -- which is equivalent to the two dereferences needed for `object` and `function` in your case, except you've lost the fast cases for booleans and numbers.

Of course, this is only if the argument is completely unknown.  If the function is inlined with a known type this compiles out to a constant in any case.

(There might be a special weirdness about `typeof val === 'function'` -- when looking at generated code, v8 doesn't seem to do the same type-tracking of function types that it does for object types.  I don't know if that actually comes into play in this case or not.)